modules/dcache/src/test/java/org/dcache/cells/MessageReplyTest.java

```diff
@@ -0,0 +1,422 @@
+        assertTrue(envelope.isReply());
+
+        assertEquals(envelope.getMessageObject(), message);
+    }
+
+    @Test
+    public void shouldAlwaysReturnFalseIfCancel() {
+        MessageReply messageReply = new MessageReply();
+        assertFalse(messageReply.cancel(false));
+        assertFalse(messageReply.cancel(true));
+    }
+
+    @Test
+    public void shouldReturnAMessageIfAMessageExists()
+            throws NoSuchFieldException, IllegalAccessException, ExecutionException, InterruptedException {
+
+        MessageReply<Message> messageReply = new MessageReply<>();
+        Message message                    = new Message();
+
+        Field messageField = messageReply.getClass().getDeclaredField("_msg");
+        messageField.setAccessible(true);
+        messageField.set(messageReply, message);
+
+        assertEquals(message, messageReply.get());
+    }
+
+    @Test (expected = ExecutionException.class)
+    public void shouldThrowAnExecutionExceptionIfAExceptionOccursDuringTheRequestForAMessage()
+            throws NoSuchFieldException, IllegalAccessException, ExecutionException, InterruptedException {
+
+        MessageReply messageReply = new MessageReply();
+        Message message           = new Message();
+        Throwable e               = new Exception();
+
+        Field messageField = messageReply.getClass().getDeclaredField("_msg");
+        messageField.setAccessible(true);
+        messageField.set(messageReply, message);
+
+        message.setReply(1, e);
+        messageReply.get();
+    }
+
+    @Test (expected = ExecutionException.class)
+    public void shouldThrowAnExecutionExceptionIfAThrowableOccursDuringTheRequestForAMessage()
+            throws NoSuchFieldException, IllegalAccessException, ExecutionException, InterruptedException {
+
+        MessageReply messageReply = new MessageReply();
+        Message message           = new Message();
+        Throwable e               = new Throwable();
+
+        Field messageField = messageReply.getClass().getDeclaredField("_msg");
+        messageField.setAccessible(true);
+        messageField.set(messageReply, message);
+
+        message.setReply(1, e);
+        messageReply.get();
+    }
+
+    @Test
+    public void shouldWaitUntilAMessageIsSet()
+            throws NoSuchFieldException, IllegalAccessException, InterruptedException {
+
+        MessageReply<Message> messageReply = new MessageReply<>();
+        Message message                    = new Message();
+
+        Thread thread = new Thread(() -> {
+            try {
+                messageReply.get();
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
+            }
+        });
+        thread.start();
+
+        TimeUnit.MILLISECONDS.sleep(100);
+        assertEquals(Thread.State.WAITING, thread.getState());
+
+        Field messageField = messageReply.getClass().getDeclaredField("_msg");
```

Comment:
Why you  don't simply `reply(message)`? The test checks that external notiffyAll will wake up the thread, but is it happens in the code?

Reply:
Yes, I agree with you and have changed the tests accordingly.